ArduCopter/mode_poshold.cpp

```diff
@@ -217,24 +215,28 @@ void ModePosHold::run()
                     brake.angle_max_roll_cd = fabsf(brake.roll_cd);
                 } else {
                     // braking angle has started decreasing so re-estimate braking time
-                    brake.timeout_roll = 1+(uint16_t)(LOOP_RATE_FACTOR*15L*(int32_t)(fabsf(brake.roll_cd))/(10L*(int32_t)g.poshold_brake_rate));  // the 1.2 (12/10) factor has to be tuned in flight, here it means 120% of the "normal" time.
+                    brake.start_time_roll_ms = now_ms;
                     brake.time_updated_roll = true;
                 }
-            }
-
-            // if velocity is very low reduce braking time to 0.5seconds
-            if ((fabsf(vel_right_cms) <= POSHOLD_SPEED_0) && (brake.timeout_roll > 50*LOOP_RATE_FACTOR)) {
-                brake.timeout_roll = 50*LOOP_RATE_FACTOR;
-            }
-
-            // reduce braking timer
-            if (brake.timeout_roll > 0) {
-                brake.timeout_roll--;
             } else {
-                // indicate that we are ready to move to Loiter.
-                // Loiter will only actually be engaged once both roll_mode and pitch_mode are changed to RPMode::BRAKE_READY_TO_LOITER
-                //  logic for engaging loiter is handled below the roll and pitch mode switch statements
-                roll_mode = RPMode::BRAKE_READY_TO_LOITER;
+                // scaling factors:
+                // 1.5 times the time taken to level the aircraft in ms
+                // 1000 to convert from seconds to ms
+                // 0.01 to convert angle_max_roll_cd to degrees
+                const uint32_t brake_timeout_roll_ms = MIN(POSHOLD_BRAKE_TIME_ESTIMATE_MAX_MS, (1.5 * 1000 * 0.01) * brake.angle_max_roll_cd / g.poshold_brake_rate_degs);
+
+                // if velocity is very low reduce braking time to 0.5seconds
+                if ((fabsf(vel_right_cms) <= POSHOLD_SPEED_0) && (now_ms - brake.start_time_roll_ms > 500) && (brake_timeout_roll_ms > 500)) {
+                    brake.start_time_roll_ms = now_ms - brake_timeout_roll_ms + 500;
+                }
+
+                // reduce braking timer
```

Comment:
I guess this comment is out-of-date?  I don't think we are "reducing braking timer" here.  It looks instead as if we are checking if sufficient time has elapsed